### ISSUE TITLE
PegNet Addresses

### ISSUE THREAD

**Issue by WhoSoup** [pegnet/pegnet#72]
So I had a look at the addresses used for balances and found that they're very inconsistent. The motivation was that you can't accidentally send pFCT to a pUSD address and vice versa for all addresses.

At the moment, the following scheme is used: 
1. take the RCD of an FA address
2. take the prefix (pFCT, pUSD, etc) to get `prefix_rcdaddress`, calculate the checksum from that
3. return `prefix_ + base58(rcd | checksum)`

The problem (imo) is that this creates wildly inconsistent addresses that range from 42 to 56 character length, as well as the fact that the prefix and _ are not base58 encoded but the rest of the address is. Regular FCT addresses fix this by using a two-byte prefix that locks the addresses inside a specific range.

So I figured why not use the same approach with PegNet addresses? Using a 4 byte prefix, I have come up with the following table:

```
        prefix  minimum address                                           maximum address
pPNT = 10715270 pPNT1k37XPNTkQLEVux7DugBBsGs7MVFrkTScSo25sdL2sWJWYzcV5 to pPNT3gnHkqQKB5ZNKVf1oXg8WU3nDyCuZbLEFWuGEZXKgsto8cy4dU
pUSD = 10791442 pUSD2bHWVHaZgM9j4jptSvmxKZuerYNRyhUcN37wGFXve6rzxS47P5 to pUSD4Y2gijcR72NrtKXo2YmueAgZyA65gYMQ17EBQwRvJ7FVaW2ZXU
pEUR = 1063b207 pEUR2HcgKZPnVpi9AZvdcpwxY5uBPPJP8ghuhPkyuRruVafuPLZtJB to pEUR4EMrZ1RdvVwGz9dYCSwurgg6W122qXahLTsE47ku9b4Q1QYLSa
pJPY = 1069b15f pJPY2EkvqtvkFPQKgptiWFu3yU4iotuwKheH9h9iuHSu1i1ST5xsxj to pJPY4BW75Lxbg4dTWQbd5su1J4qdvWdb2YX4nmFy3yLtfiPw59wL78
pGBP = 10664f09 pGBP11bzrGaLGA4bfdK8qWYJU7wh22DdwhN2ijKMz991zkpLhuvf4s to pGBP2xMB5icBgqHjVD23R8YFniic8dwHeYEpMoRc8q31emCqKyu7DG
pCAD = 106026c9 pCAD2rtN7pdXABrSHpbhgMvxrAAK6YcXY6mS4nZrNuwGJcF5QnF8H5 to pCAD4odYMGfNas5a7QJcFyvvAkwEDALBEweDhrg6XbqFxcda2rDaRU
pCHF = 10605656 pCHF1aipkPMuD1amas35xrEdWb3rd36ra5t9Tu3bnMQv81ADSh6SAB to pCHF3XTzyqPkdgouQSjzYUEaqBpmjepWGvkw6y9qw3Jun1Yi4m4tJa
* pINR is not possible since Base58 doesn't contain uppercase i
pSGD = 1075c07e pSGD1zpbPnLidnWw18xiwJStxaLQXiCuCuKRyEqZ4rysr6EannPHaf to pSGD3wZmdENa4Tk4pifdWvSrHB7KeKvYukCDcJwoDYssW6d5QrMjj4
pCNY = 10607a1e pCNY1RckjdsiN5yr7yT3iEwUxYeDU2wWu4hz6dAHsEtUZFaVnCRYxX to pCNY3NMvy5uZnmCywZ9xHrwSH9R8aefAbuamjhGY1vnUDFxzQGQ16v
pHKD = 10680c09 pHKD2c2pCaUjxLgsZTQrUuokZrh3RJhrcBocvjKWAGdxNV7eE13AvB to pHKD4YmzS2WbP1v1P37m4XohtTTxXvRWK2gQZoRkJxXx2VW8r51d4a
pXAU = 107d4217 pXAU1h3c5XXQgP4cxgEyrneAkvuBv3jEqJ1URTDVXyNBsPTy4NbZTV to pXAU3dnnJyZG74HknFwtSQe85Xg72fStY8tG4XKjgfGBXPrTgSa1bt
pXAG = 107d40b1 pXAG1nYT4r6LNFzMGdkSLavhxHfM5aeWjcXmc6g9uiSKGzdJRrABXM to pXAG3jHdJJ8BnwDV6DTLvCvfGtSGCCNASTQZFAnQ4QLJw11o3v8dfk
pXPD = 107d9839 pXPD21XUx9fbwCZhzVnKWh5ZoCdh55wL6TnDj4zrRW58xpsDJudC9d to pXPD3xGfBbhTMsnqp5VE6K5X7oQcBheyoJf1N976aBy8cqFhvybeJ2
pXPT = 107d99db pXPT2YhDsm1omuE4uQbFp8kFqpLUkW74Sho8tfzq32nescyVFYTcnb to pXPT4VSQ7D3fCaTCizJAPkkDAR7Ps7pi9YfvXk75BigeXdMyscS4vz
pXBT = 107d48bc pXBT2DJgVyqwNoGyQAw1sZHbEnAtwaE1ehTxcvLUL3E5Wm9NruALDM to pXBT4A3rjRsnoUW7DkdvTBHYZNwp4BwfMYLkFzSiUj85AmXsUy8nMk
pETH = 1063aa56 pETH2orKkpWsn9oesLDeFiWv7Z63npMPbSaYCwncy1kcER4EvTAJfR to pETH4kbVzGYjCq2nguvYqLWsS9rxuS53JHTKr1ts7hebtRSjYX8kop
pLTC = 106cda3e pLTC1CjfhqVzPJFWL626hEyNvhr1aJa2SzSi1oHAFB6D6HzwrUwioH to pLTC39UqwHXqoyUe9fj1GryLFJcvgvHg9qKVesPQPrzCkJPSUYvAwg
pXBC = 107d46fc pXBC1MoecuYAmopZspxtBVeHaBEdqgWnBBBSHC7BDscjwmQJGD8PDy to pXBC3JYprMa2CV3hhQfnm7eEtn1YxJERt24DvGDRNZWjbmnntH6qNN
pFCT = 1064ce2e pFCT1B2PhYMgHFHMvPVWKf1aunnHtfPac164YAsW99Ga5ymA8bYknj to pFCT37mZvzPXhvWVjyCQuH1YEPZD1H7EJqxrBEykHqAZjz9ekfXCw8
```

Unfortunately the Indian rupee doesn't work with this approach. Using all lowercase characters means that LTC no longer works. If there's a possibility of dropping or renaming INR (ie RUP) then this would be a very clean approach to giving a unique, human readable address that has a consistent length.

Another possibility is taking the current system and just adding a prefix to stabilize the length. Those prefixes could be something like:
```
PEG (32 byte payload + 4 byte checksum) = 22 0b 70
PEG (37 byte payload + 4 byte checksum) = 44 5d e6
```

Which would make the addresses: 
`pUSD_PEG1k2fvhGErEKCpVDHMjBY5bdMBbdmkA4VDDsknZBwUcPPkJmCS7yAQ6tR5`
5 characters non base58 name + 60 characters base58(3 byte prefix + 5 byte name + 32 byte rcd + 4 byte checksum)

The FA address `FA1y5ZGuHSLmf2TqNf6hVMkPiNGyQpQDTFJvDLRkKQaoPo4bmbgu` (all zeros for RCD) would be: 
```
pPNT_PEG1k2RTjAL59eNe9ekjpcysXGgFtcFjeSw6PHCi22br1oVmuNG9cJ6BmEZE
pUSD_PEG1k2fvhGErEKCpVDHMjBY5bdMBbdmkA4VDDsknZBwUcPPkJmCS7yAQ6tR5
pEUR_PEG1k1sqf63fidaQPfzco36Vx28WZsjXG6zsrJaTK8ybvQ8AX5PbUpV2UJ7K
pJPY_PEG1k28C9tukGsENbEDRYn3tz9RoZMEtPPs3oFDGrWvZvZ3xEVKAehvCdoMK
pGBP_PEG1k1yPYWxXkQKYYwLirpaMikb3hPN6EQ46mVsEHmLwFPSQn4esD1CCtV1u
pCAD_PEG1k1mrGNmLD1TZb91cLZB5sabcPv4TapvQvRkESmEu2vDCAycp4UdjaiQF
pCHF_PEG1k1mvqkzXvXuPevcGUtwA1Whic9nE4KLgZACNVJnisqm33RfhCKVKnvFf
pINR_PEG1k25HgVmwTa9Gio2VyKDNZ72Htp9qEcGn1L7wp5q9T7eMeRuSbTPJVX3V
pSGD_PEG1k2a2ZaahfU6tffRrZivbsRhizZAyeifqet3XMkhXXZszRMTHJBLTueFD
pCNY_PEG1k1mzone2zVEDsyF4jZXdyLUPYntb3xk3fsqmdGTPzCWFZWbU33nGGVBn
pHKD_PEG1k22NYC9JGqHaovtR7ktpg8ourWBgdYqwfMiRxTuSpD7rvXHXV9VFC829
pXAU_PEG1k2pNRzofa17ZoQGqcAd2ruuL9DQXZ87YEsyWrTQRsyaVM3ymCdtMEe2X
pXAG_PEG1k2pNPvgqZk4yMFSuxvSH7rNpzn5i53JQZ2BLaGgc3cjAUMRoG17FBfK9
pXPD_PEG1k2pYBXXFMygd56mTtTFh6asQeb5WzDANggbdCMti4a6RUQBne7Man9Up
pXPT_PEG1k2pYDtoLWZT1sZRfUjTywoBhPNbJyANp3NwFoRro1qdwL4FB9QVGseVY
pXBT_PEG1k2pP5ifTN9zxSbtu67zutEZ49bXj1vLt7b7tN7FfZovdZ5AG13arQ96p
pETH_PEG1k1spyjowdhj7pnJ3bPaAqs8xoaZy4XZMnzT61dGw5ZxyunvGamefMrtU
pLTC_PEG1k2E11NJfKeSC2ZcKd5FHwW2gEA1TjKp7L1dm5Ugij5PjjNzzdHrnXeza
pXBC_PEG1k2pP3CojeQsfTzKZXKmrUwqscdvSnXvnvUWXbbfT4632CwMADREW3LsE
pFCT_PEG1k1vX4MWvKpwD4FquNU4E6jfDCRySx5xzHa1C4UX3fovZgmLbVceciRUz
```

They would still be uniquely identifiable even without the 5 character name at the front, since the name is encoded inside the base58 itself, allowing for the potential of just having "PEG....." addresses. 

Thoughts?

**Comment by WhoSoup** [comment]
Published the code to find prefixes if anyone wants to play around with it: https://github.com/whosoup/base58-prefix

**Comment by MitchellBerry** [comment]
The first approach is much neater but hamstrung with any future additions that fall outside of base58 along with INR currently. Rupee is the name of a bunch of different currencies, sticking to international standards is probably a good idea.

Plain evil offhand suggestion: utf8 currency symbols perhaps?

**Comment by WhoSoup** [comment]
> Plain evil offhand suggestion: utf8 currency symbols perhaps?

Don't think those would fit anywhere where they wouldn't make things more difficult for both us and end-users.

**Comment by MitchellBerry** [comment]
Agree, thought to throw it out there though. 
On the 2nd approach, it's really deep into the address before you can actually do some basic visual checking. Can always check from the end instead but many web apps and exchanges cutoff addresses visually.

Base62 would certainly solve a few problems, doubt that's a popular idea with many.  A static size and clean look for addresses is certainly ideal. The first approach fits this but also means remapping anything with either I or O in the ticker.

**Comment by MitchellBerry** [comment]
Quite sure there aren't any variable length address formats anywhere else.

**Comment by PaulBernier** [comment]
Thinking more about it I am in favor of dropping those `pX` prefixes and have a single universal address starting with `PEGxxx`. Why? This `pX` prefix is purely a UX feature, and we are forcing the second layer (wallet) to adopt this UX feature for "more security". I am on the side of leaving the wallets decide how to best handle their user experience. This pX_ prefix is not very elegant, and will be redundant in many cases. I imagine if you have a UI you would have selected the asset you want to transact anyway, so why add this extra prefix to repeat the same information?

Following this logic we could even go as far as keeping the original FAxxx address, but I think it's not a bad thing to use PEG instead and take a bit of distance with Factom as if the PegNet was blockchain independent.

**Comment by PaulBernier** [comment]
This scheme is assuming people will say： "send me 5 at pUSD_PEGxxx" while in reality what happens is "send me 5 pUSD to PEGxxx", and there is no need to re-iterate that you want pUSD.

Also having a single address across all assets is a better UX advantage. With different addresses for each asset it becomes impossible for the user to get even remotely familiar with his own address. I don't know my credit card number but I am familiar with its 4 last digits, which is useful, sometimes.

**Comment by PaulBernier** [comment]
Sharing @PaulSnow position from Discord (and my answer inline with the above):

```
Paul Snow:

This discussion about a token type + address or encoded address(token type + address) boils down to passing around two keys to find a balance and to build a transaction, or passing around one key to find a balance and build a transaction.
If we use the PegNet as (primarily) a payment solution, asking for pUSD_68GHhgu7678hju and being paid from pUSD_8757857hklHGKgkkjghgy7 is certain to send USD to a USD request.
Asking for pUSD kwelrj43jlkqljgv4qjaglkjqer and sending from □ pUSD □ pEUR □ pJPY □ ... kwelrj43jlkqljgv4qjaglkjqer is asking for issues, especially if all FAT tokens are options. Of course, the wallet can be smarter and present the user with the right asset. In fact both approaches of handling the two keys independently and handling them as a single encoded address are in fact functionally the same, and anyone building infrastructure can handle them either way without anyone else being the wiser.

Paul Bernier:

Yes, but the 2-key solution offer more flexibility over the 1-key solution (of course we can start parsing the 1 key and split it to recover both pieces of information, which is not very clean). I think Who GitHub issue highlighted that the 1-key solution doesn't have a natural elegant solution, while the 2 key is just more flexible (more freedom to upper layers to handle addresses). That's why I'm personally more inclined towards the later.
```

**Comment by sambarnes** [comment]
I'm in agreement with @PaulBernier here.

My thoughts from discord:
> If we're trying to make integration to exchanges (and addition of new assets) as seamless as possible, wouldn't it be easier to just hold balances of all pegged assets in the same address, and leaving asset type as part of the transaction header? kinda like Ethereum does with all ERC-20, ERC-721, ERC-du-jour tokens
> hmm I imagine under this situation, wallets would just work around the issue by reformatting the human-readable address to a different prefix for ease of use. Since the actual on-chain address will just be the RCD hash like we do for FCT right? So the same RCD hash can be used across assets, the wallet could  just strip the prefix, add the correct prefix for the asset being sent
> of course we can always allow for separate address formats if a user wants to be extra careful about things (so that the token type checkbox must also match the address being sent to) but yea to me it seems like a UX thing for wallets to worry about down the line, and the daemon just have universal addresses under the hood

**Comment by WhoSoup** [comment]
> Since the actual on-chain address will just be the RCD hash like we do for FCT right?

Yeah, it's actually the **same** RCD hash as FCT addresses. 

Using `22 0b 70` as prefix, they would be functionally identical to FA addresses apart from the 3-byte prefix opposed to 2-bytes.

I agree with @PaulBernier that it makes more sense to leave this kind of error checking to wallets than the protocol itself. Something like a convention of `pUSD@PEG.................` that can be copy-pasted into wallets and pre-fill with the relevant settings, but doesn't require a separate address for every asset.

That still leaves the question of whether or not to support **optional** "asset-only" addresses at the protocol level or not. It's possible to just encode the asset into the checksum (which is what's done at the moment), which means just the last 4 bytes change. 

It could also be possible to create multi-currency addresses by using an additional 3 bytes to create a 24-bit bitmap of which assets are accepted. (we have 20 assets atm)

**Comment by PaulSnow** [comment]
I am having trouble understanding how a 1 key solution is less flexible than a 2 key solution.  What I am looking at in Factom is all the problems we have with ack/msg code where we don't tie them together in transmission where nobody cares that there are two components, but must tie them together before we can process them.

Where you have one key, you can break them into two keys and do whatever you want.  But when you are asking for something or transmitting something, and in particular where code is signing something, there is no question about what you are managing/transmitting/requesting/ whatever.

I don't feel strongly here, but experience tends to tell me when you are communicating across systems, managing two keys is more problematic than handling one key.  In this case, the components don't travel around separately (like Acks and Msgs in Factomd) but when I first wrote the code in PegNet with two keys, there was more code and more parameters than when I went back through the code and reduced it to 1 key.

**Comment by PaulSnow** [comment]
@WhoSoup You are right, we can move to a convention where we use @ instead of _, so instead of:
```
    CAD   pCAD_29gz1WiDWF76vtocEVBnrYn9aSt1tToMgdVFwkPVdRN5jvgXqa
    CHF   pCHF_29gz1WiDWF76vtocEVBnrYn9aSt1tToMgdVFwkPVdRN5mcahSb
    CNY   pCNY_29gz1WiDWF76vtocEVBnrYn9aSt1tToMgdVFwkPVdRN5gQgBPm
    ETH   pETH_29gz1WiDWF76vtocEVBnrYn9aSt1tToMgdVFwkPVdRN5iW77Wt
    EUR   pEUR_29gz1WiDWF76vtocEVBnrYn9aSt1tToMgdVFwkPVdRN5hKZLHH
    FCT   pFCT_29gz1WiDWF76vtocEVBnrYn9aSt1tToMgdVFwkPVdRN5kTxkdt
    GBP   pGBP_29gz1WiDWF76vtocEVBnrYn9aSt1tToMgdVFwkPVdRN5gDUQPy
    HKD   pHKD_29gz1WiDWF76vtocEVBnrYn9aSt1tToMgdVFwkPVdRN5iGUuxq
    INR   pINR_29gz1WiDWF76vtocEVBnrYn9aSt1tToMgdVFwkPVdRN5hqfZyk
```
we would have
```
    CAD   pCAD@29gz1WiDWF76vtocEVBnrYn9aSt1tToMgdVFwkPVdRN5jvgXqa
    CHF   pCHF@29gz1WiDWF76vtocEVBnrYn9aSt1tToMgdVFwkPVdRN5mcahSb
    CNY   pCNY@29gz1WiDWF76vtocEVBnrYn9aSt1tToMgdVFwkPVdRN5gQgBPm
    ETH   pETH@29gz1WiDWF76vtocEVBnrYn9aSt1tToMgdVFwkPVdRN5iW77Wt
    EUR   pEUR@29gz1WiDWF76vtocEVBnrYn9aSt1tToMgdVFwkPVdRN5hKZLHH
    FCT   pFCT@29gz1WiDWF76vtocEVBnrYn9aSt1tToMgdVFwkPVdRN5kTxkdt
    GBP   pGBP@29gz1WiDWF76vtocEVBnrYn9aSt1tToMgdVFwkPVdRN5gDUQPy
    HKD   pHKD@29gz1WiDWF76vtocEVBnrYn9aSt1tToMgdVFwkPVdRN5iGUuxq
    INR   pINR@29gz1WiDWF76vtocEVBnrYn9aSt1tToMgdVFwkPVdRN5hqfZyk
```
Note that the checksums are different in the addresses because it includes the type of asset.  Maybe we don't care about that?

Another question.  Does breaking up the address in two keys mean that all FAT tokens get mixed in with the PegNet tokens in the wallet?  Do we have some benefits from this?  Or risk of confusion?  I guess its nothing we cannot address in the Wallets and their UI...

**Comment by PaulSnow** [comment]
By the way, Bitcoin does have three address lengths for the 3 address types. 

I am not entirely sure what having a fixed address length gives us.  Addresses are so long, the user can't visually tell the lengths apart.  Most people don't use fixed length characters without making the effort to do so, and thus the addresses above generally look to them like:

    CAD   pCAD_29gz1WiDWF76vtocEVBnrYn9aSt1tToMgdVFwkPVdRN5jvgXqa
    CHF   pCHF_29gz1WiDWF76vtocEVBnrYn9aSt1tToMgdVFwkPVdRN5mcahSb
    CNY   pCNY_29gz1WiDWF76vtocEVBnrYn9aSt1tToMgdVFwkPVdRN5gQgBPm
    ETH   pETH_29gz1WiDWF76vtocEVBnrYn9aSt1tToMgdVFwkPVdRN5iW77Wt
    EUR   pEUR_29gz1WiDWF76vtocEVBnrYn9aSt1tToMgdVFwkPVdRN5hKZLHH
    FCT   pFCT_29gz1WiDWF76vtocEVBnrYn9aSt1tToMgdVFwkPVdRN5kTxkdt
    GBP   pGBP_29gz1WiDWF76vtocEVBnrYn9aSt1tToMgdVFwkPVdRN5gDUQPy
    HKD   pHKD_29gz1WiDWF76vtocEVBnrYn9aSt1tToMgdVFwkPVdRN5iGUuxq
    INR   pINR_29gz1WiDWF76vtocEVBnrYn9aSt1tToMgdVFwkPVdRN5hqfZyk



And the decode of the address validates them.  The checksum reduces the chance of a typo or short address losing funds to 1 in ~4 billion, and those odds don't materialy change with fixed length addresses.

Is there a fundamental gain that we get by having fixed length human readable addresses?

**Comment by MitchellBerry** [comment]
Could be minor OCD, but fixed length *seems* right despite the reality that monospace fonts aren't that common for general usage. 

To keep throwing things out from leftfield: 
Mapping keys to a standardised word set like the eff's largewordlist: https://www.eff.org/deeplinks/2016/07/new-wordlists-random-passphrases
The average person is quite averse to base58 encoding from my experience. 

Comes across as the same problem that occurs with config markdowns, either you prioritise humans or computers, there's no clean in-between.

**Comment by WhoSoup** [comment]
> I am having trouble understanding how a 1 key solution is less flexible than a 2 key solution. What I am looking at in Factom is all the problems we have with ack/msg code where we don't tie them together in transmission where nobody cares that there are two components, but must tie them together before we can process them.

I'm not sure what you mean by a two key solution. It's more or less just a formatting change. With the current addresses you still need two components: 1) the value, 2) the address. You can send "5" to "pUSD...." and the system knows what is meant. If you have generic addresses, you'll have to send "5 pUSD" to "generic address".

This is also about which convention to use at the protocol level, not necessarily at the end-user level. For example, we can use the generic PEG addresses but wallets can add an additional layer providing currency-specific addresses (which we can require as a wallet convention). The wallet would then be responsible for converting the currency-address to the right format.

This way they'd have the option of doing both. Having twenty different addresses in the database at protocol level even though every single address is functionally the same address seems like unnecessary clutter.

As for the fixed length:
1. It's easier to document and debug. It's a minor gain but being able to say "all addresses are xx bytes in length" is easier than "they're between 42 and 56 bytes".
2. You can distinguish addresses if you end up running multiple pegnet networks or want to future proof them. If we ever add another address type for any reason, we can give it a new prefix and not have to worry about accidental collisions

**Comment by PaulSnow** [comment]
Note:  Explanation below is mostly for readers, as I know I am not telling anyone else in this discussion something they don't know:

To make base58 one length, you just add a character to prevent the fact that base58 is really conversion of the hex number (the address) to a base58 number, so you need a leading digit or two to ensure it is the same magnitude in base58.

Thus fixed length is always longer, so I prefer shorter and rely on the checksum to ensure digits are correct, as the length is too long for anyone to use it as a measure of correctness anyway.

But not religious about that.

When it comes to the two key vs one key discussion, I think that using the 1 key in the protocol that encodes the type makes sense because you will have to carry around the type no matter what you do anyway.   And one field is easier, with the type is easier to debug, and limits some error cases (both human and computer)

Once you talk about how the wallets organize the tokens, then conversions to the underlying key are always there, conversions to the keys are always there, and what buckets a wallet uses to hold balances are not impacted either way (just when you apply the type to the address).   Having written the code both ways, I lean a little more to a one key solution that embeds the type, but if I lose the consensus on the matter I am not going to cry about it.

**Comment by WhoSoup** [comment]
I spent some time reading into the FAT documentation, trying to figure out a best way to approach integration, and FAT is using plain FA..... addresses for everything. There's no standard that really suits PegNet so we're going to have to write our own, which means we can easily use any addresses we want, but that's another consideration.

Another consideration is exchanges. What's more likely going to be integrated with exchanges? Having to track 20 asset-addresses for every address might drive some exchanges away. Or it might not! Maybe there will be pUSD-only exchanges. I don't know the answer to this one.

Right now I'm leaning toward something simple at the protocol level (ie fatd / pegnet miners), and then a standard of asset-specific addresses at the wallet level with the option of them using generic addresses while being explained about the additional risks. That gives people the choice of security but also simplifies the core code.

The way I see it work is that, for example, someone posts a pegnet address on a website, and person A copy/pastes it into their wallet.
If it's a generic address:
The wallet presents them with a choice of currencies they have to choose from. if there's a "asset@...." in front of the address, it will auto-fill that choice and if they pick a different one, it will display a warning.

If it's an asset-specific address:
The wallet displays only that asset but there is a button that lets them change currency. If they pick another currency, the wallet will display a warning that they're sending undesired currency. Then they have the choice of a) converting the asset they want to send to the other one on the fly and send the right one, or b) they want to send a different asset and the wallet automatically converts the asset-address to the other one

Thoughts?

**Comment by PaulSnow** [comment]
Let's decide what we are doing here and just do it.

**Comment by PaulSnow** [comment]
My preference is to make each asset independent at the transaction points.  So from the wallet, you can select the assets that you are holding (i.e. 
```
PNT_lkargwj    113.46
USD_IOUH8778   345.27
CHY_qerfgw243 1024.333
```
Now a user doing a conversion gets to do something like
```
     Source Asset ______    amount _______
Destination Asset ______    amount _______
```
For the Source, they are given only the assets where they have a balance.  And maybe they pick the asset first, then the address holding the asset second.
For the Destination, they are only given the assets, then it fills in the destionation address.
Then they get to pick either a source amount or a destination amount.
The Wallet looks to the side and warns them if they don't have enough source assets to cover the desitionation amount if they are low.  The other way just works.

Now when a user is doing a transaction, they get to do this:

```
     Source Asset ______    amount _______
Destination Asset ______    amount _______
```
And they fill it all out, but the source and destination have to match.  We could allow them not to match, but if we do then we have to submit both a conversion and a transaction.

I do believe (for scalability) we should only handle one input.  We can pay multiple outputs, but specifying one input speeds transaction processing should we need to shard for greater transaction volume.

Now internally, we can do what we want.  Use a field and a address, or use the combined human encoding, or whatever.  But treating each token as its own thing has both user reasons and marketing reasons.  We want people to think of these assets as independent tokens that happen to self enforce their pegs rather than a single asset.

**Comment by PaulSnow** [comment]
If we are willing to go down this route, then the given address encodings are good enough?
Thoughts?

**Comment by drkatz** [comment]
Hey everyone. I think @WhoSoup has a good point about the complexity requiring both end users and exchanges to not only track but understand the 20+ asset address prefixes. I expect the asset list will also dynamically change over time as new pairs are added. This becomes an issue when there needs to be 100% determinism about addresses validation and formatting, yet this is a dynamic system.

The idea with the design of FAT addresses was we decided that regardless of how you prefix an address, what you are doing is really the same despite the asset. We built FAT to have FAxxx... addresses to be "baskets" for assets, a single address can hold any number of different types of FAT tokens. In practice there is no confusion caused by this homogenous address structure as when you ask for a balance from the API, you always do it for a FAxxx address in the context of a specific token.

I do like the idea of using address formats to query specific pegnet balances from an address, however, I think this can also be implemented with less confusion for end users and exchanges. For example instead of submitting a request to the `get-balance` endpoint with pUSD_xxx as the address parameter, you could submit FAxxx as the address parameter and also include a "pair" parameter set to "USD" that denotes that the API should get's the pUSD balance. I think any abstraction like this should be in the API and not protocol level

**Comment by WhoSoup** [comment]
> This becomes an issue when there needs to be 100% determinism about addresses validation and formatting, yet this is a dynamic system.

Can you elaborate on this? The multiple-address system is deterministic, it's just `asset prefix | rcd hash of FA address | checksum over asset + rcd`. You can take an asset-specific address and easily convert it to any other asset address (or the generic one).

I think the question is which one will fit easier into the FAT spec: one generic address that manages 20+ tokens with the ability to convert between the tokens, or an address that manages a single token with the ability to send/convert tokens to a different address type.

**Comment by PaulSnow** [comment]
> I think the question is which one will fit easier into the FAT spec: one generic address that manages 20+ tokens with the ability to convert between the tokens, or an address that manages a single token with the ability to send/convert tokens to a different address type.

I'd phrase it differently, because what we have done is create one address that handles all the tokens, either way.

The question is:  Do we expose the underlying address on a day to day basis with the user and the exchanges?  Or do we treat the tokens as independent tokens in the wallets from a user/exchange point of view?

I really am on the outside here because I didn't look at the FAT discussion and there is inertia in that direction.  So we need to get this done.  I believe the fact that the address holds many tokens should absolutely not be exposed where it isn't needed, and I think exchanges don't need it and for the most part users don't need it.  Where anyone cares (on the back end) anyone can treat it as a token with many assets.  But it is only on the back end that anyone would care, and they can do it.  On the front end, I see no reason for the user to look in their wallet and see "baskets of assets".  I see every reason that they will view their wallet as having a "basket of assets".  That's baskets vs basket.  A Collection of baskets of assets is more confusing than a simple basket of assets.

And to put a fine point on this, on the back end, where anyone codes the support for the PegNet, neither way of encoding user facing addresses (or really, if we encode the asset with the address vs we have a tag separate from an address) makes no difference.  It is the difference of sending two parameters to a function or a struct with two fields.  I can handle either with little issue.

**Comment by sambarnes** [comment]
I honestly don't get why we are still discussing this. Its something we **easily** punt on for wallet UI/UX months down the line, and just use factoid RCD hash as the on chain universal address for now. Most everyone is in agreement on this it seems, especially considering the vote that took place in discord where everyone opted for single universal addresses in the protocol, with asset specific prefixes being an optional wallet convention.

> On the front end, I see no reason for the user to look in their wallet and see "baskets of assets". I see every reason that they will view their wallet as having a "basket of assets". That's baskets vs basket. A Collection of baskets of assets is more confusing than a simple basket of assets.

One factoid address, many assets. That's a single basket of assets. If they want the option of having displaying different addresses for every single asset they hold, so be it. Just a UX thing down the line, we don't even have any user testing to show that its the case that people don't like ethereum's single address style, so we might as well go with what people are already comfortable with for now

**Comment by drkatz** [comment]
@WhoSoup I would agree the multi address system is deterministic from the point of view of the system, it matters less from that perspective. In this case I mean determinism from the point of view of a user or exchange on the address and it's contents. Visually they can have a single Factoid address that holds all the Factom ecosystem tokens: Factoids, and all FAT tokens including pegged tokens (excluding EC sorry :open_mouth: ). Perhaps determinism is the wrong word. From an API perspective I feel it's easier to ask an API ask "which varieties of tokens does my single address own" than "which addresses do I own, which tokens do those own".

The single address multi asset approach certainly fits better with FAT as an existing system as FAT-0 and FAT-1 using that design with the express purpose to begin to form a basket of Facotm assets. Forgive me if this sounds far fetched, but I like the idea of a FAxxx address being a "Factom Address" that holds Factom assets, it's good branding in my opinion.

Good points @sambarnes, I think can and should take cues here from successful systems like Ethereum, basically all which use a single address format to hold assets (contract based tokens). Exchanges, wallet users, existing wallets, and wallet engineers are already primed for this approach. I genuinely believe it will be easier to get them to understand this mechanism than prefixed addresses for specific assets, as they're already used to the concept of a "token address" (the erc-20 token's contract address) which is a analog of our token chain ID.

I think if people want this feature that is totally acceptable, but it should be a wallet feature and be outside of the protocol.

**Comment by PaulSnow** [comment]
The backend guys can view 1) the prefixed address or 2) a tag + FAxxxx address or 3) a tag + RCD any way they like, because all forms 1) 2) and 3) are trivially converted into one another on the backend.

On the front end, these addresses are not trivally converted.  So users and developers using tools to look at the data structures in the blockchain and in transactions can't easily convert addresses without tools and infrastructure, which might not be handy.

User Experience:
Ethereum is not a good example.  

If you look at the "balance" of a set of contracts (these tokens cannot be transferred) then all of these [tokens are lost](https://etherscan.io/token/0xa74476443119A942dE498590Fe1f2454d7D4aC0d).

As of this writing, in the last hour, looking at GNT in Golem, 12136.40 tokens have been lost.  Who knows the value of these tokens, but that is a scary number of lost tokens.

What is wild is that this came from a [warning](https://www.reddit.com/r/ethereum/comments/60ql37/attention_be_careful_using_ethereum_tokens/) about Ethereum tokens published on Reddit 2 years ago, and tokens continue to be lost.

Miner Experience:
Miners are the lifeblood of cryptocurrency security.  And these people need to configure their mining, then look at and validate their mining.  This is facilitated by clarity where the data structures they will be looking at are understandable.  So the address (1 2 or 3) they look at in the config file needs to be the address we put in the OPR, and that needs to be the address that they look at to check their balances.

Backend Experience
At any point in building backend code, we can convert 1, 2, and 3 into whatever form is best for what we are doing.  Anywhere humans may be typing or cutting and pasting addresses, 3 is inappropriate.  Losing tokens due to typos or even (this happened to me once) having cut and paste not work accurately is just not a good user experience.  1 and 2 are checksummed making the risk of a typo losing funds is over 1 in 4 billion.

The ability of an exchange to request funds like an address is a single asset address avoids warnings like:

> IMPORTANT Send only BTC to this deposit address.
> Sending any other currency to this address may result in the loss of your deposit!
> BTC deposits will be available after 1 network confirmations.

Note that this warning is replicated on deposits on Poloniex. Coinbase has such warnings too, but I found XRP interesting: "If you send funds without an XRP Tag or with an incorrect XRP Tag, your funds cannot be credited to your account"

All these warnings point to a problem users have _and a real problem for exchanges_.  It costs them time, money, and the good will of their clients when these mistakes occur.

Why exactly are we saying that doing addressing the way other protocols do it is good for exchanges instead of giving them a feature that avoids these problems AND lets them code their backends however they like?

Where any request for funds is made that may involve the user typing or cutting and pasting addresses, 2 is inappropriate.  It is very easy to get the tag confused relative to the address, resulting in the wrong token type being sent to a requested address.  While not quite as bad as sending tokens into the bit void created by non-checksummed RCD addresses, the user is still in a position of requesting the sender to return the inappropriate tokens for the intended tokens.  In the PegNet where auto conversions are possible, the refund might be more confusing until new infrastructure comes about to help.

**Comment by WhoSoup** [comment]
> The backend guys can view 1) the prefixed address or 2) a tag + FAxxxx address or 3) a tag + RCD any way they like, because all forms 1) 2) and 3) are trivially converted into one another on the backend.

My vote is (still) 2. I agree with your assessment that 3) is just not viable in any way and I don't think having them use the raw rcd hash was ever on the table. The choice is between using FA addresses that have the asset baked in, and FA addresses with the asset sent separately. 

I'd be okay with miners also being able to put in their pPNT address into the config and having the software convert it to FA on the fly. Requiring people to use pPNT addresses in the OPR is kind of redundant since the mining can only ever pay out PNT and nothing else. 

I also think treating each asset-address as a separate address is misleading. Understanding that one address holds many assets should be one of the first things people using pegnet learn. In the normal cryptospace, it's well known that you can't tell if two separate addresses belong to the same person, but in pegnet, separate asset-addresses use the same private key to sign a transaction. 

You can also (if I understand it right) only convert assets using a single address. This is going to lead to some confusion if wallets end up using BIP44 generation.

Example:
User makes 5 pUSD addresses, which generates `FA1...FA5`. They receive some pUSD on FA5 and convert it to pEUR. Now the wallet would have to display that pEUR address as the fifth pEUR address even though it's the "first" pEUR address the user made. If they list all five pEUR addresses it's going to cause a lot of UI clutter.

If that isn't pointed out well enough, they might import their seed phrase on a new wallet and then generate the "first" pEUR address (which uses FA1) and wonder what happened to their money or if their seed phrase is wrong since it displays a different pEUR address. 

**tl;dr** users of pegnet should absolutely be aware that multiple asset-addresses belong to a single FA address. given that assumption, i don't see any reason to use anything but FA addresses in the OPRs / conversions / transactions formats

**Comment by WhoSoup** [comment]
> The ability of an exchange to request funds like an address is a single asset address avoids warnings like:
> 
> > IMPORTANT Send only BTC to this deposit address.
> > Sending any other currency to this address may result in the loss of your deposit!
> > BTC deposits will be available after 1 network confirmations.
> 
> Note that this warning is replicated on deposits on Poloniex. Coinbase has such warnings too, but I found XRP interesting: "If you send funds without an XRP Tag or with an incorrect XRP Tag, your funds cannot be credited to your account"

I agree but this is something we can solve with the asset-address convention at the wallet level. They tell their wallet to create a pUSD only address for FA.... and then publish that. If someone copy/pastes that into a wallet, the wallet will know to only send pUSD to that address. If a wallet doesn't understand the asset-address convention, they will be unable to send funds at all. 

The possibility of a **bad** wallet allowing someone to send pEUR to a pUSD address exists no matter which of our options we pick.

**Comment by sambarnes** [comment]
I don't think anyone on either side is going to change their mind at this point. 
 
Anyways, I figured #fct-market channel was a fair representation of the typical target user for pegnet, so I posted a poll over the weekend:
![poll](https://user-images.githubusercontent.com/1211977/61640924-c5b19480-ac63-11e9-9251-3038216568ff.png)

Small sample size, but I think the user feedback speaks loudly enough. Keep universal as the default in the protocol (likely just `FA...` address) like users have been trained to expect, and make typed addresses later on as a UX _experiment_ for wallet implementations

**Comment by Cahl-Dee** [comment]
Assigning to @PaulSnow as he has committed to taking the lead on this ticket and driving it to completion. If a call is required, it will be conducted.

**Comment by PaulSnow** [comment]
The compromise is to take the address in the Config File for the miner and put it in the OPR address.  As long as we can decode it, we can make it work for past records, and for new records if people would rather use a different scheme for addresses.

As we have a single address encoding already, I'd move we use it, and we can replace it later with something else.

**Comment by Cahl-Dee** [comment]
@PaulSnow will pull out the PNT address functionality and we will go with the use of Factoid addresses across the board. This means the FCT address will go in the OPR entry and config file.

The CLI rewards command will also take a FCT address now.

Assigning this Issue to him as well.

**Comment by PaulSnow** [comment]
While I am in the OPR record, I'd like to make it conform to the FAT standard for our project.  The downside to this is that it is in a pull request to a FAT repository and is far from optimal for collaboration.  So the changes we are looking at:

Removal of the Oracle Price Record Chain ID
rename the FactomDigitalIdentity to minerid
Change the address for the coinbase to a [miner confusing] FAxxxxx address
Compression of the winners to a 8 leading bytes of the hash of all 10, concatenated with the first byte of all 10 (makes reconstruction of the 10 without the grading algorithm possible for debugging and light weight verification)

**Comment by drkatz** [comment]
Thanks @PaulSnow, most of those changes mentioned were made to the spec today so please check out the latest!

There's a conversation ongoing in the standard PR about the winner hashes so I'll continue there as it's not relevant to this issue. I have some questions about the shorthash & concat approach.

**Comment by PaulSnow** [comment]
Against my sense of a more understandable view of the PegNet, the reward address is no longer the PNT_xxxx format address, but a FAxxxxx address.

The FAxxxxx address is now the address in the OPR record.

Because the tPNT_xxxx and PNT_xxx format is really useful in the PegNet code, the FA address is pulled from the config file, and converted into the PNT_xxx format within the OPR.  But the PNT_xxx is not part of the JSON marshalling, so the FAxxxxx address goes out to the OPR record.

This preserves the addressing, balance checking, management of the Test Network code and Main Network code, balances and addresses.  Striping all of all of this code from the code base would have been a huge engineering task, and all we would have gained is having to start from scratch with testing, and a loss of coded and check summed addresses (I believe) we will add back when this hits actual payment use cases and the miners.

---------------
Also fixed as part of address shuffling was the fact that PNT is the token on the main net for PegNet, not pPNT.

Also fixed as part of address shuffling is the rework of the naming used for the asset address conversion functions.  Made raw always []byte, made raw (for the RCD or hash of public key in the case of ECs) always referred to as raw.  Made everything "Covert<whatever>To<whatever>.

**Comment by PaulSnow** [comment]
> Thanks @PaulSnow, most of those changes mentioned were made to the spec today so please check out the latest!
> 
> There's a conversation ongoing in the standard PR about the winner hashes so I'll continue there as it's not relevant to this issue. I have some questions about the shorthash & concat approach.

The winner hash changes discussed are no longer on the table.

Carl closed the door on new features for PegNet a number of days ago. That didn't make the wire.

**Comment by PaulSnow** [comment]
BTW, I moved this back to "To Do" simply because after having backed out PNT_xxxx addresses from the config and the OPRs, I know of nothing else literally in process.  But I am not sure we are finished with this, so someone else needs to make the call to close this out, or move it to milestone2.

**Comment by Cahl-Dee** [comment]
Internal discussion has suggested any further work on this can be pushed off to pM3. Happy to discuss further if others disagree.
